gordon_janitor_gcp/gce_authority.py

```diff
@@ -0,0 +1,181 @@
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        _metadata_blacklist = self.config.get('metadata_blacklist', [])
+        metadata_blacklist = [dict([pair]) for pair in _metadata_blacklist]
+        keyfile = self._get_or_raise('gce_keyfile')
+        scopes = self.config.get('gce_scopes')
+
+        gce_auth = auth.GoogleAuthClient(
+            keyfile, scopes=scopes, session=self.session)
+        return gcp_clients.GCEClient(
+            gce_auth,
+            self.session,
+            tag_blacklist=tag_blacklist,
+            metadata_blacklist=metadata_blacklist)
+
+    def build_authority(self):
+        """Coordinate dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
+        """
+        crm_client = self._get_crm_client()
+        gce_client = self._get_gce_client()
+        msg_gen = self._get_msg_generator()
+
+        return GCEAuthority(self.config, crm_client, gce_client, msg_gen,
+                            self.rrset_channel, **self.kwargs)
+
+
+def get_authority(config, rrset_channel, **kwargs):
+    builder = GCEAuthorityBuilder(config, rrset_channel, **kwargs)
+    return builder.build_authority()
+
+
+@zope.interface.implementer(interfaces.IAuthority)
+class GCEAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        crm_client (obj): client used to fetch GCE projects.
+        gce_client (obj): client used to fetch instances for each project.
+        msg_generator (obj): obj used to transform instance data into resource
+            record set messages for the reconciler.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+    def __init__(self,
+                 config,
+                 crm_client,
+                 gce_client,
+                 msg_generator,
+                 rrset_channel=None,
+                 **kwargs):
+        self.config = config
+        self.crm_client = crm_client
+        self.gce_client = gce_client
+        self.msg_generator = msg_generator
+        self.rrset_channel = rrset_channel
+
+    async def _get_active_project_ids(self):
+        active_projects = await self.crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def _get_projects(self):
+        projects = await self._get_active_project_ids()
+        project_blacklist = set(self.config.get('project_blacklist', []))
+        # TODO: emit a metric for all projects and a metric for
+        # project - blacklist.
+        return sorted(projects - project_blacklist)
+
+    async def _get_instances(self, projects):
+        for project in projects:
+            yield await self.gce_client.list_instances(project)
+
+    async def start(self):
+        """Gather instance data batches and send them to the
```

Comment:
This needs to be one single line [per the style guide](https://google.github.io/styleguide/pyguide.html#Comments).